pkg/cleanup/cleanup.go

```diff
@@ -0,0 +1,235 @@
+package cleanup
+
+import (
+	"context"
+	"database/sql"
+	"fmt"
+	_ "github.com/go-sql-driver/mysql"
+	"github.com/icinga/icingadb/pkg/icingadb"
+	"github.com/icinga/icingadb/pkg/utils"
+	"go.uber.org/zap"
+	"sync"
+	"time"
+)
+
+// Tables type defines mapping of history tables with corresponding period.
+type Tables map[string]string
+
+type Options struct {
+	Interval time.Duration `yaml:"interval"  default:"1h"`
+	Count    uint          `yaml:"count"     default:"5000"`
+}
+
+// tableconfig is defined for table configuration, with table name, period and start time.
+type tableconfig struct {
+	Table     string
+	Period    time.Duration
+	Starttime time.Time
+}
+
+// Cleanup is a cleanup wrapper with icingadb.DB wrapper and,
+// logging and wait group synchronization capabilities.
+type Cleanup struct {
+	HistoryTables Tables
+	logger        *zap.SugaredLogger
+	db            *icingadb.DB
+	wg            *sync.WaitGroup
+	opts          *Options
+}
+
+// NewCleanup returns a new cleanup.Cleanup wrapper for the configured history tables.
+func NewCleanup(db *icingadb.DB, tables Tables, opts *Options, logger *zap.SugaredLogger) *Cleanup {
+	return &Cleanup{
+		HistoryTables: tables,
+		logger:        logger,
+		db:            db,
+		wg:            &sync.WaitGroup{},
+		opts:          opts,
+	}
+}
+
+// cleanupFunc type is declared to define cleanup functions signature.
+type cleanupFunc func(ctx context.Context, db *icingadb.DB, tbl tableconfig, limit uint) (sql.Result, error)
+
+// cleanupFuncs type is declared to define mapping of history tables with their corresponding cleanupFunc.
+type cleanupFuncs map[string]cleanupFunc
+
+// Start starts the cleanup routine for the configured history tables.
+func (c *Cleanup) Start() error {
+	if c.HistoryTables != nil {
+		// cufuncConfig maps the history tables to corresponding cleanup methods.
+		var cufuncConfig = func() cleanupFuncs {
+			funcConfig := make(cleanupFuncs)
+
+			for table, cufunc := range map[string]cleanupFunc{
+				"acknowledgement": cleanAckFunc,
+				"comment":         cleanCommentFunc,
+				"downtime":        cleanDowntimeFunc,
+				"flapping":        cleanFlappingFunc,
+				"notification":    cleanNotificationFunc,
+				"state":           cleanStateFunc,
+			} {
+				funcConfig[table] = cufunc
+			}
+			return funcConfig
+		}()
+
+		ctx := context.Background()
+		for table, period := range c.HistoryTables {
+			tempPeriod, _ := time.ParseDuration(period)
+			c.wg.Add(1)
+			go c.controller(ctx, tableconfig{table + "_history", tempPeriod, time.Now()}, cufuncConfig[table])
+		}
+
+		c.wg.Wait()
+	}
+
+	return nil
+}
+
+// controller calls the cleanup method periodically after c.opts.Interval.
+func (c *Cleanup) controller(ctx context.Context, tbl tableconfig, cufunc cleanupFunc) {
```

Comment:
This function should return any errors instead of logging them. Also you should reset the timer. A generic function for that could look like this:

```
func StablePeriodic(ctx context.Context, tick time.Duration, callback func(ctx context.Context, time time.Time) error) (context.CancelFunc, <-chan error) {
	ctx, cancelCtx := context.WithCancel(ctx)
	g, ctx := errgroup.WithContext(ctx)
	ticker := time.NewTicker(tick)

	g.Go(func() error {
		defer ticker.Stop()

		for {
			select {
			case t := <-ticker.C:
				if err := callback(ctx, t); err != nil {
					return err
				}

				ticker.Reset(tick)
			case <-ctx.Done():
				return ctx.Err()
			}
		}
	})

	return cancelCtx, com.WaitAsync(g)
}